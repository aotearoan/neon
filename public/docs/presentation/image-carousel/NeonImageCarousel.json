{
  "displayName": "NeonImageCarousel",
  "description": "<p>\n  The <strong>NeonImageCarousel</strong> is used to display a carousel of images for the user to swipe\n  through. NOTE: It is recommended to explicitly set the width & height of the carousel via CSS, this will\n  ensure it will not be impacted by images with different aspect ratios. Images are automatically resized to fit the\n  carousel dimensions.\n</p>",
  "tags": {},
  "exportName": "default",
  "props": [
    {
      "name": "images",
      "description": "The images to be displayed in the carousel.",
      "type": {
        "name": "Array as () => Array<NeonCarouselImage>"
      },
      "required": true
    },
    {
      "name": "imageCountLabel",
      "description": "Provide an alternative image count label. This can be useful for translations. The default is e.g.\n<em>2 images</em>.",
      "type": {
        "name": "string"
      },
      "defaultValue": {
        "func": false,
        "value": "undefined"
      }
    },
    {
      "name": "previousLabel",
      "description": "Provide an alternative label for the Previous button.",
      "type": {
        "name": "string"
      },
      "defaultValue": {
        "func": false,
        "value": "'Previous'"
      }
    },
    {
      "name": "nextLabel",
      "description": "Provide an alternative label for the Next button.",
      "type": {
        "name": "string"
      },
      "defaultValue": {
        "func": false,
        "value": "'Next'"
      }
    }
  ],
  "events": [
    {
      "name": "current-image",
      "description": "The index of the currently visible image.",
      "type": {
        "names": [
          "number"
        ]
      }
    }
  ]
}