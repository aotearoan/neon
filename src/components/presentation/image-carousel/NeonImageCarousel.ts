import { defineComponent, onMounted, onUnmounted, ref } from 'vue';
import type { NeonCarouselImage } from '@/common/models/NeonCarouselImage';
import NeonLink from '@/components/navigation/link/NeonLink.vue';
import NeonButton from '@/components/user-input/button/NeonButton.vue';

/**
 * <p>
 *   The <strong>NeonImageCarousel</strong> is used to display a carousel of images for the user to swipe
 *   through. NOTE: It is recommended to explicitly set the width & height of the carousel via CSS, this will
 *   ensure it will not be impacted by images with different aspect ratios. Images are automatically resized to fit the
 *   carousel dimensions.
 * </p>
 */
export default defineComponent({
  name: 'NeonImageCarousel',
  components: {
    NeonButton,
    NeonLink,
  },
  props: {
    /**
     * The images to be displayed in the carousel.
     */
    images: { type: Array as () => Array<NeonCarouselImage>, required: true },
    /**
     * Provide an alternative image count label. This can be useful for translations. The default is e.g.
     * <em>2 images</em>.
     */
    imageCountLabel: { type: String, default: undefined },
    /**
     * Provide an alternative label for the Previous button.
     */
    previousLabel: { type: String, default: 'Previous' },
    /**
     * Provide an alternative label for the Next button.
     */
    nextLabel: { type: String, default: 'Next' },
  },
  emits: [
    /**
     * The index of the currently visible image.
     *
     * @type {number}
     */
    'current-image',
  ],
  setup(props, { emit }) {
    const initialised = ref<boolean>(false);
    const currentImage = ref<number>(0);
    const carouselItems = ref<HTMLUListElement | null>(null);
    const carouselItem = ref<Array<HTMLLIElement>>([]);
    const observers = ref<Array<IntersectionObserver>>([]);

    const changeImage = (index: number) => {
      if (index !== currentImage.value) {
        currentImage.value = index;
        emit('current-image', index);
      }
    };

    const scrollTo = (index: number) => {
      if (carouselItems.value && props.images.length > 0) {
        changeImage(index);
        carouselItems.value.scrollTo(carouselItems.value.clientWidth * index, 0);
      }
    };

    const next = () => {
      if (currentImage.value < props.images.length - 1) {
        scrollTo(currentImage.value + 1);
      }
    };

    const previous = () => {
      if (currentImage.value !== 0) {
        scrollTo(currentImage.value - 1);
      }
    };

    onMounted(() => {
      const options = {
        root: carouselItems.value,
        rootMargin: '0px',
        threshold: 0.6,
      };

      observers.value = carouselItem.value.map((el, index) => {
        const obs = new IntersectionObserver((entries) => {
          entries.forEach((entry) => {
            // filter out the observations on initialization
            if (initialised.value && entry.isIntersecting) {
              changeImage(index);
            }
          });
        }, options);
        obs.observe(el);

        return obs;
      });

      setTimeout(() => {
        // initialise scroll position to first element
        if (carouselItems.value && carouselItems.value.scrollLeft !== 0) {
          carouselItems.value.scrollLeft = 0;
        }
        initialised.value = true;
      }, 50);
    });

    onUnmounted(() => {
      observers.value.forEach((observer) => observer.disconnect());
    });

    return {
      emit,
      currentImage,
      carouselItem,
      carouselItems,
      initialised,
      next,
      previous,
      scrollTo,
    };
  },
});
