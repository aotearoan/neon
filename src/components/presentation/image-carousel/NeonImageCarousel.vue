<template>
  <div
    :class="{ 'neon-image-carousel--initialised': initialised }"
    class="neon-image-carousel"
    tabindex="0"
    @keydown.stop.prevent.left="previous"
    @keydown.stop.prevent.right="next"
  >
    <div class="neon-image-carousel__container" tabindex="-1">
      <neon-button
        :circular="true"
        :disabled="currentImage === 0"
        :title="previousLabel"
        :transparent="true"
        button-style="text"
        class="neon-image-carousel__previous"
        color="neutral"
        icon="chevron-left"
        size="l"
        @click="previous"
      />
      <ul ref="carouselItems" class="no-style neon-image-carousel__items">
        <li
          v-for="(image, index) in images"
          :key="image.src"
          ref="carouselItem"
          :class="{ 'neon-image-carousel__item--active': index === currentImage }"
          class="neon-image-carousel__item"
        >
          <img :alt="image.alt" :src="image.src" class="neon-image-carousel__image" />
        </li>
      </ul>
      <neon-button
        :circular="true"
        :disabled="currentImage === images.length - 1"
        :title="nextLabel"
        :transparent="true"
        button-style="text"
        class="neon-image-carousel__next"
        color="neutral"
        icon="chevron-right"
        size="l"
        @click="next"
      />
    </div>
    <div class="neon-image-carousel__nav" tabindex="-1">
      <neon-link
        v-for="(_image, index) in images"
        :key="index"
        :aria-label="`Display image ${index + 1}`"
        class="neon-image-carousel__nav-item-link"
        outline-style="none"
        role="button"
        tabindex="0"
        @keydown.stop.prevent.enter="scrollTo(index)"
        @keydown.stop.prevent.space="scrollTo(index)"
      >
        <div
          :class="{ 'neon-image-carousel__nav-item--active': index === currentImage }"
          class="neon-image-carousel__nav-item"
          tabindex="-1"
          @click="scrollTo(index)"
        >
          <div class="neon-image-carousel__nav-item-indicator"></div>
        </div>
      </neon-link>
    </div>
    <span class="neon-image-carousel__label" tabindex="-1">
      {{ imageCountLabel || `${images.length} ${images.length === 1 ? 'image' : 'images'}` }}
    </span>
  </div>
</template>

<script lang="ts" src="./NeonImageCarousel.ts" />
