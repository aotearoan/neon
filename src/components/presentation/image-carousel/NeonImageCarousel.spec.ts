import type { RenderResult } from '@testing-library/vue';
import { fireEvent, render } from '@testing-library/vue';
import NeonImageCarousel from './NeonImageCarousel.vue';
import { router } from '@/../test/unit/test-router';

describe('NeonImageCarousel', () => {
  const images = [
    {
      src: '/images/image-carousel/pic-01.jpg',
      alt: 'Pic 1',
    },
    {
      src: '/images/image-carousel/pic-02.jpg',
      alt: 'Pic 2',
    },
  ];

  let harness: RenderResult;
  let scrollToMock: jest.Mock;

  beforeEach(() => {
    scrollToMock = jest.fn();
    HTMLUListElement.prototype.scrollTo = scrollToMock;

    harness = render(NeonImageCarousel, {
      props: { images },
      global: {
        plugins: [router],
      },
    });
  });

  it(`renders carousel`, () => {
    const { html } = harness;
    expect(html()).toMatchSnapshot();
  });

  it(`renders previous disabled when first element`, () => {
    const { getByTitle } = harness;
    expect((getByTitle('Previous') as HTMLButtonElement).disabled).toEqual(true);
    expect((getByTitle('Next') as HTMLButtonElement).disabled).toEqual(false);
  });

  it(`renders next disabled when last element`, async () => {
    const { getByTitle } = harness;
    await fireEvent.click(getByTitle('Next'));
    expect((getByTitle('Previous') as HTMLButtonElement).disabled).toEqual(false);
    expect((getByTitle('Next') as HTMLButtonElement).disabled).toEqual(true);
  });

  it(`updates image styles when next is clicked`, async () => {
    const { container, getByTitle } = harness;
    // check image item styles
    expect(container.querySelectorAll('.neon-image-carousel__item')[0].className).toMatch(
      'neon-image-carousel__item--active',
    );
    expect(container.querySelectorAll('.neon-image-carousel__item')[1].className).not.toMatch(
      'neon-image-carousel__item--active',
    );

    await fireEvent.click(getByTitle('Next'));

    // check image item styles
    expect(container.querySelectorAll('.neon-image-carousel__item')[0].className).not.toMatch(
      'neon-image-carousel__item--active',
    );
    expect(container.querySelectorAll('.neon-image-carousel__item')[1].className).toMatch(
      'neon-image-carousel__item--active',
    );
  });

  it(`updates nav styles when next is clicked`, async () => {
    const { container, getByTitle } = harness;
    // check nav buttons
    expect(container.querySelectorAll('.neon-image-carousel__nav-item')[0].className).toMatch(
      'neon-image-carousel__nav-item--active',
    );
    expect(container.querySelectorAll('.neon-image-carousel__nav-item')[1].className).not.toMatch(
      'neon-image-carousel__nav-item--active',
    );

    await fireEvent.click(getByTitle('Next'));

    // check nav buttons
    expect(container.querySelectorAll('.neon-image-carousel__nav-item')[0].className).not.toMatch(
      'neon-image-carousel__nav-item--active',
    );
    expect(container.querySelectorAll('.neon-image-carousel__nav-item')[1].className).toMatch(
      'neon-image-carousel__nav-item--active',
    );
  });

  it(`checks image styles when previous is clicked`, async () => {
    const { container, getByTitle } = harness;
    await fireEvent.click(getByTitle('Next'));
    await fireEvent.click(getByTitle('Previous'));

    // check image styles
    expect(container.querySelectorAll('.neon-image-carousel__item')[0].className).toMatch(
      'neon-image-carousel__item--active',
    );
    expect(container.querySelectorAll('.neon-image-carousel__item')[1].className).not.toMatch(
      'neon-image-carousel__item--active',
    );
  });

  it(`checks nav styles when previous is clicked`, async () => {
    const { container, getByTitle } = harness;
    await fireEvent.click(getByTitle('Next'));
    await fireEvent.click(getByTitle('Previous'));

    // check nav buttons
    expect(container.querySelectorAll('.neon-image-carousel__nav-item')[0].className).toMatch(
      'neon-image-carousel__nav-item--active',
    );
    expect(container.querySelectorAll('.neon-image-carousel__nav-item')[1].className).not.toMatch(
      'neon-image-carousel__nav-item--active',
    );
  });

  it(`navigates to image when a nav button is clicked`, async () => {
    const { container } = harness;

    await fireEvent.click(container.querySelectorAll('.neon-image-carousel__nav-item')[1]);

    // check image styles
    expect(container.querySelectorAll('.neon-image-carousel__item')[0].className).not.toMatch(
      'neon-image-carousel__item--active',
    );
    expect(container.querySelectorAll('.neon-image-carousel__item')[1].className).toMatch(
      'neon-image-carousel__item--active',
    );

    // check nav buttons
    expect(container.querySelectorAll('.neon-image-carousel__nav-item')[0].className).not.toMatch(
      'neon-image-carousel__nav-item--active',
    );
    expect(container.querySelectorAll('.neon-image-carousel__nav-item')[1].className).toMatch(
      'neon-image-carousel__nav-item--active',
    );
  });

  it('emits current-image event', async () => {
    const { emitted, getByTitle } = harness;
    await fireEvent.click(getByTitle('Next'));
    expect(emitted()['current-image']).toEqual([[1]]);
  });
});
