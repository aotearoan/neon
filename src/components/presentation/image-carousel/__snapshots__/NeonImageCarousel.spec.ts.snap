// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonImageCarousel renders carousel 1`] = `
<div class="neon-image-carousel" tabindex="0">
  <div class="neon-image-carousel__container" tabindex="-1"><button class="neon-button--l neon-button--neutral neon-button--text neon-button--state-ready neon-button--text-transparent neon-button--disabled neon-button--circular neon-button--with-icon neon-button--icon-only neon-button neon-image-carousel__previous neon-image-carousel__previous" disabled="" tabindex="-1" title="Previous">
      <div class="neon-icon--name-chevron-left neon-icon--neutral neon-icon--disabled neon-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" width="100%" height="100%">
          <path class="neon-svg--fill" d="M15.28 5.22a.75.75 0 0 1 0 1.06L9.56 12l5.72 5.72a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215l-6.25-6.25a.75.75 0 0 1 0-1.06l6.25-6.25a.75.75 0 0 1 1.06 0Z"></path>
        </svg>
      </div>
      <!--v-if-->
      <!--v-if-->
    </button>
    <ul class="no-style neon-image-carousel__items">
      <li class="neon-image-carousel__item--active neon-image-carousel__item"><img alt="Pic 1" src="/images/image-carousel/pic-01.jpg" class="neon-image-carousel__image"></li>
      <li class="neon-image-carousel__item"><img alt="Pic 2" src="/images/image-carousel/pic-02.jpg" class="neon-image-carousel__image"></li>
    </ul><button class="neon-button--l neon-button--neutral neon-button--text neon-button--state-ready neon-button--text-transparent neon-button--circular neon-button--with-icon neon-button--icon-only neon-button neon-image-carousel__next neon-image-carousel__next" tabindex="0" title="Next">
      <div class="neon-icon--name-chevron-right neon-icon--neutral neon-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" width="100%" height="100%">
          <path class="neon-svg--fill" fill-rule="evenodd" d="M8.72 18.78a.75.75 0 001.06 0l6.25-6.25a.75.75 0 000-1.06L9.78 5.22a.75.75 0 00-1.06 1.06L14.44 12l-5.72 5.72a.75.75 0 000 1.06z"></path>
        </svg>
      </div>
      <!--v-if-->
      <!--v-if-->
    </button>
  </div>
  <div class="neon-image-carousel__nav" tabindex="-1"><a class="neon-link--outline-none neon-link neon-image-carousel__nav-item-link neon-image-carousel__nav-item-link" tabindex="0" aria-label="Display image 1" role="button"><span class="neon-link__label"><!-- @slot the content of the link --><div class="neon-image-carousel__nav-item--active neon-image-carousel__nav-item" tabindex="-1"><div class="neon-image-carousel__nav-item-indicator"></div></div></span></a><a class="neon-link--outline-none neon-link neon-image-carousel__nav-item-link neon-image-carousel__nav-item-link" tabindex="0" aria-label="Display image 2" role="button"><span class="neon-link__label"><!-- @slot the content of the link --><div class="neon-image-carousel__nav-item" tabindex="-1"><div class="neon-image-carousel__nav-item-indicator"></div></div></span></a></div><span class="neon-image-carousel__label" tabindex="-1">2 images</span>
</div>
`;
